### ISSUE TITLE
forkproxy segmentation fault when add proxy device

### ISSUE THREAD

**Issue by m-shibata** [lxc/lxd#6112]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails:
```
config: {}                                                          
api_extensions:                                                                            
- storage_zfs_remove_snapshots                                      
- container_host_shutdown_timeout                                   
- container_stop_priority                                           
- container_syscall_filtering                                       
- auth_pki                                                          
- container_last_used_at                                            
- etag                                                              
- patch                                                             
- usb_devices                                                       
- https_allowed_credentials                                         
- image_compression_algorithm                                       
- directory_manipulation                                            
- container_cpu_time                                                
- storage_zfs_use_refquota                                          
- storage_lvm_mount_options                                         
- network                                                           
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- candid_authentication
- candid_config
- candid_config_key
- usb_optional_vendorid
- id_map_current
api_status: stable
api_version: "1.0"
auth: trusted                                                                                                                                                                                        [6/299]
public: false                                                       
auth_methods:                                                       
- tls                                                               
environment:                                                        
  addresses: []                                                     
  architectures:                                                    
  - x86_64                                                          
  - i686                                                            
  certificate: |                                                    
    -----BEGIN CERTIFICATE-----                                     
    MIIFRjCCAy6gAwIBAgIQTwIHYepDly2Bw5QmT6mdnzANBgkqhkiG9w0BAQsFADA2
    MRwwGgYDVQQKExNsaW51eGNvbnRhaW5lcnMub3JnMRYwFAYDVQQDDA1yb290QHVi
    dW50dXZiMB4XDTE5MDgyMjExMDYzMFoXDTI5MDgxOTExMDYzMFowNjEcMBoGA1UE
    ChMTbGludXhjb250YWluZXJzLm9yZzEWMBQGA1UEAwwNcm9vdEB1YnVudHV2YjCC
    AiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoCggIBAKlvfP3jPD7FeDAR+N16PFCK                       
    9C8gZF6KWH5GCpSUtp2H6oaS4gviJgphGH1bu4Jvi5+RVfF+lookJvcKZZb/nLeP
    odNO6EPVtZ+4UlaaYmokOfKRU6Gw8rFM5+U0C2apxXuZa9UGe8iOZPiXWWCax2LC
    vM5aJ2A2zQ96MMsGn8XSYyiscLqWdTUsBeA4gSTzzwXCcr+nWdw/9YBW6MmXox9B
    EieWtjvma413Os+nC5fa263ueRuLOnjJDjHwhESPCn2fMWLoxsvMs399N0HPMIwj
    ZIHSF/14Qeaw9TKLiKH3n6Et3CZ5i+ui36xAkljNzj4WC57M4C0iXydpCFmPFH5L
    1mmle8HGXf5hieZYQPyCW/oj/x3C0WN1DbYRnjb8NyOAL2Hmdl6MPTT4WHK1l2FA
    xS6RngrXosUwUoFU5o8DdS4T+2Vyu1wNdwAheJBYgXesYZ6n+OK4dKil8LpeeZ/k
    ZtE3XhklnlFFGWQwRjY/re1GAx2bxBVxP0J3M9/XnOg/ABlWkLi0uId+ZPtqF6ml
    4qBhmmc55io2f8MeJYbWXwvrvgjarpT4LRKigK+LvE03uYzAp491zxpzNfJQkOfw
    KixYhO9GP7TqGSLAZGv4HMDtEeTc+9pmYh1wilxr5HhDYq3vMEOMn6Z8CHB3+XwN
    4xcf4aPgbHOLDyfvQEbtAgMBAAGjUDBOMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUE
    DDAKBggrBgEFBQcDATAMBgNVHRMBAf8EAjAAMBkGA1UdEQQSMBCCCHVidW50dXZi
    hwTAqHpgMA0GCSqGSIb3DQEBCwUAA4ICAQBNrA3EqBv9C7pPmid6bI92y1Lh+w5y
    J8KcUpeOMh+VgO7iMXvjxyCkPWQsqH/TJ8PlUSYYDNA1ihenaSEwL2PqXcoR63s1
    lnTLjacnoBI3i6DYi8oEyp79IDGoWpJQW6uFkLpwToxnLLvoxL8U4ADOcTzltY0Y
    Aeujte4g0vXRqPamu7x3JN175atrb655OHw31y3KjpalKWSdQ88xeo3f6jOfq7bR
    XTTIL3wBQl6DIChrr044n5HHkE/idzBSeqhzbGJduJ0Zbk8DW5rtYRBBLM+caBtp
    48hD1gv+uzOTu9xZCoJ+uT+cmWzwjiTlt9V/VtdN0Co6ru1wyAr68RReM4lwL5h8
    DjNqX5zUiqJjbLmbYQ3NMQ/8Kmx3SXwNBxcqVJuvWLy/uCsY4uTLVfmP3NgZyFpx
    Ts0fezoUqiC2J+iQHOACIh0FkIv3gmn8Q7R3Y4qe1Tw3f0x0N4Azs9feM83gNxVi
    21Gs8mwaCyBCVcwmu/vjRHeFAk5RhPLlFm0uT/5V8L5dBWTlfjnf+deHeHkMgKpg
    kyIcpzCouSqiKNRP9jy7+bWvN/IURoKyvvn9aqAZfgEBaKEdDCev1T5SilAREYPR
    A0kIRzlu+OoMQVeOSSViFU3Yp3HUIT3JbEOW2Y+DcIB7B/iiNyvfqrKNQq/8Wk+e
    k/WvAc8FNeRpyg==
    -----END CERTIFICATE-----
  certificate_fingerprint: fe4f8c2e74e27cdf4fb44e2139179a1e0035b94db817699bf5fc9cba8ae36223
  driver: lxc
  driver_version: 3.0.4
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features: {}
  kernel_version: 4.15.0-58-generic
  lxc_features: {}
  project: ""
  server: lxd
  server_clustered: false
  server_name: ubuntuvb
  server_pid: 5009
  server_version: 3.0.4
  storage: btrfs
  storage_version: "4.4"
```

# Issue description

When proxy device is added, `lxc config device add` succeed but forkproxy segmentation fault.

 - Ubuntu 18.04 LTS + lxd 3.0.3-0ubuntu1~18.04.1 (apt)
  => doesn't occur
 - Ubuntu 18.04 LTS + lxd 3.0.4 (snap)
  => occur
 - Ubuntu 18.04 LTS + lxd 3.16 (snap)
  => doesn't occur

# Steps to reproduce

 1. `$ lxc launch ubuntu:18.04 n0`
 2. `$ lxc config device add n0 port80 proxy listen=tcp:0.0.0.0:80 connect=tcp:127.0.0.1:80`
 3. `$ sudo cat /var/snap/lxd/common/lxd/logs/n0/proxy.port80.log`

```
$ sudo cat /var/snap/lxd/common/lxd/logs/n0/proxy.port80.log
Error: remove 0.0.0.0:80: read-only file system
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5a3398]

goroutine 1 [running]:
os.(*File).Name(...)
        /snap/go/4098/src/os/file.go:50
net.FileListener(0x0, 0x0, 0x0, 0x0, 0x0)
        /snap/go/4098/src/net/file.go:36 +0x88
main.(*cmdForkproxy).Run(0xc000010740, 0xc000243680, 0xc00007bda0, 0x6, 0x6, 0x0, 0x0)
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/main_forkproxy.go:352 +0x2de
github.com/spf13/cobra.(*Command).execute(0xc000243680, 0xc00007bc80, 0x6, 0x6, 0xc000243680, 0xc00007bc80)
        /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:826 +0x465
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001daa00, 0xc0001c3f60, 0x1, 0x1)
        /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:914 +0x2fc
github.com/spf13/cobra.(*Command).Execute(...)
        /build/lxd/parts/lxd/go/src/github.com/spf13/cobra/command.go:864
main.main()
        /build/lxd/parts/lxd/go/src/github.com/lxc/lxd/lxd/main.go:164 +0xd32
```

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [x] Container log (`lxc info NAME --show-log`)
```
$ lxc info n0 --show-log
Name: n0
Remote: unix://
Architecture: x86_64
Created: 2019/08/22 14:01 UTC
Status: Running
Type: persistent
Profiles: default
Pid: 9854
Ips:
  eth0: inet    10.69.24.84
  eth0: inet6   fd42:f197:b5e4:24a9:216:3eff:fec5:9075
  eth0: inet6   fe80::216:3eff:fec5:9075
  lo:   inet    127.0.0.1
  lo:   inet6   ::1
Resources:
  Processes: 24
  CPU usage:
    CPU usage (in seconds): 3
  Memory usage:
    Memory (current): 153.76MB
    Memory (peak): 206.61MB
  Network usage:
    eth0:
      Bytes received: 14.98kB
      Bytes sent: 6.16kB
      Packets received: 69
      Packets sent: 55
    lo:
      Bytes received: 1.01kB
      Bytes sent: 1.01kB
      Packets received: 12
      Packets sent: 12

Log:

lxc n0 20190822140113.290 ERROR    start - start.c:proc_pidfd_open:1607 - Function not implemented - Failed to send signal through pidfd
```
 - [x] Container configuration (`lxc config show NAME --expanded`)
```
$ lxc config show n0 --expanded
architecture: x86_64
config:
  image.architecture: amd64
  image.description: ubuntu 18.04 LTS amd64 (release) (20190813.1)
  image.label: release
  image.os: ubuntu
  image.release: bionic
  image.serial: "20190813.1"
  image.version: "18.04"
  volatile.base_image: 2dd611e2689a8efc45807bd2a86933cf2da0ffc768f57814724a73b5db499eac
  volatile.eth0.host_name: vethIU0WP6
  volatile.eth0.hwaddr: 00:16:3e:c5:90:75
  volatile.idmap.base: "0"
  volatile.idmap.current: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":1000000,"Nsid":0,"Maprange":1000000000},{"Isuid":false,"Isgid":true,"Hostid":1000000,"Nsid":0,"Maprange":1000000000}]'
  volatile.last_state.power: RUNNING
devices:
  eth0:
    name: eth0
    nictype: bridged
    parent: lxdbr0
    type: nic
  port80:
    connect: tcp:127.0.0.1:80
    listen: tcp:0.0.0.0:80
    type: proxy
  root:
    path: /
    pool: default
    type: disk
ephemeral: false
profiles:
- default
stateful: false
description: ""
```
 - [x] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
```
t=2019-08-22T14:01:11+0000 lvl=info msg="Creating container" ephemeral=false name=n0
t=2019-08-22T14:01:11+0000 lvl=info msg="Created container" ephemeral=false name=n0
t=2019-08-22T14:01:13+0000 lvl=info msg="Starting container" action=start created=2019-08-22T14:01:11+0000 ephemeral=false name=n0 stateful=false used=1970-01-01T00:00:00+0000
t=2019-08-22T14:01:13+0000 lvl=info msg="Started container" action=start created=2019-08-22T14:01:11+0000 ephemeral=false name=n0 stateful=false used=1970-01-01T00:00:00+0000
```
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

**Comment by tomponline** [comment]
@stgraber I've re-created this issue, it only occurs in the snap. Investigating now.

**Comment by tomponline** [comment]
@stgraber So I think this is the issue:

1. This `os.Remove` (https://github.com/lxc/lxd/blob/lxd-3.0.4/lxd/main_forkproxy.go#L314) is attempted for all listener types, even TCP ones, however if it fails with `os.IsNotExist` then the error is skipped and it proceeds normally.
2. I believe this is OK on a non-snap system as the OS will always return IsNotExist when trying to delete the listen name as a file "0.0.0.0:80", as it never exists. However inside the snap environment, the error returned is not `IsNotExist`, but `read-only file system` which then means the rest of the file listener is not attempted and nil is eventually use to try and establish a listener, at which point Go crashes hard.
3. I believe this was fixed in https://github.com/lxc/lxd/commit/abdacc46a9a90bec9dabe9e8ee8eaf59b263faa3#diff-b70c0d1b1e3ce93f3e57209039f1259cR317 because it adds checks around the `os.Remove` to only be done if the listener type is unix.

**Comment by tomponline** [comment]
@stgraber I suspect this would also have been an issue in the lxd 3.0.3 snap too.

How would you like me to proceed? Should I produce a new PR against the lxc-3.0 branch with a fix to this?

**Comment by stgraber** [comment]
@tomponline yeah, a PR against stable-3.0 would be good as the commit which fixed this isn't going to be cherry-picked (no multi-port support in 3.0).

Note that you won't be able to have the PR build, that's okay. I'll review the change and merge it, then deal with getting things building later :)

**Comment by stgraber** [comment]
This got merged in 3.0 branch
